### PR TITLE
Surrogate keys

### LINKED ISSUE #23: Set better cache-control headers
*@trieloff commented on Feb 23, 2018, 11:10 AM UTC:*

*   etags (based on revision)
*   Vary (don't)
*   [Surrogate Keys](https://docs.fastly.com/guides/purging/getting-started-with-surrogate-keys) (based on template function, so that all stale content can be invalidated)

*This issue was moved by [trieloff](https://github.com/trieloff) from [adobe/project-helix#44](https://github.com/adobe/project-helix/issues/44).*

### PR DESCRIPTION
**Pull request by trieloff**
Fixes #23 

Each URL that is fetched individually to render the content should be listed in `content.sources`.

Each URL in `content.sources` will then be sha256 hashed and added to `Surrogate-Keys`.

If you want to write a selective cache invalidator (👋 @tripodsan), go over each file in the commit event, and go over each name of the commit (e.g. branch or tag names when pushed), calculate the `raw.githubusercontent.com` URL and get the hash.

### CODE REVIEW

File: src/html/set-surrogate-key.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  // somebody already set a surrogate key
+  if (!(response && response.headers && response.headers['Surrogate-Key']) && (content && content.sources && Array.isArray(content.sources))) {
+    logger.debug('Setting Surrogate-Key header');
+    return {
+      response: {
+        headers: {
+          'Surrogate-Key': content.sources.map((uri) => {
+            const hash = crypto.createHash('sha256');
+            hash.update(uri);
+            return hash.digest('hex');
```
Comment (tripodsan) [inline src/html/set-surrogate-key.js]: shoulnd't we truncate this a bit or encode with base64? the hex encoded sha256 has 64 characters, which yields to very long surrogate-keys. using base64 would make it 4 times smaller.

Comment (trieloff) [reply src/html/set-surrogate-key.js]: @tripodsan base64 sounds good, truncating not so much.

Comment (tripodsan) [reply src/html/set-surrogate-key.js]: you didn't change it to base64.... ;-)

Comment (trieloff) [reply src/html/set-surrogate-key.js]: Doesn't mean I'm not open to the change.

**Pull request merged**